Fix/issue125 - add `use_zone_redundancy` variable to set zone redundancy options for regions where it doesn't exist

johnwildes opened the pull request:
Fixes #125 

## Description
This pull request introduces several updates, focusing on enhancing zone redundancy support, improving container runtime checks, and refining documentation. The most significant changes include adding a `use_zone_redundancy` variable across multiple modules, improving the container runtime validation logic in the `avm` script, and updating the documentation to reflect these changes.

### Zone Redundancy Enhancements:
* Added a new `use_zone_redundancy` variable in `variables.tf` and `modules/container-registry/variables.tf` to control zone redundancy for container registries and app environments. Defaults to `true`. [[1]](diffhunk://#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288eR89-R94) [[2]](diffhunk://#diff-4a2d4c67b203238ccbf62e1400a59fbabbe4d1d9abe554474fbbed6a469eb782R69-R74)
* Updated `main.container.app.environment.tf`, `main.container.registry.tf`, and `main.virtual.network.tf` to use the new `use_zone_redundancy` variable for enabling zone redundancy in container apps, registries, and public IP zones. [[1]](diffhunk://#diff-6b79819cb19c9c402eabc662abd050668000b65f40dd63fa4893204d3d5514f5L12-R12) [[2]](diffhunk://#diff-b36c2e55e1f6d00aa635179d44ec700511762c73f32541cf3a398688e85f82feR15) [[3]](diffhunk://#diff-bce0d19513233e0593ba9f2b583b996dfbac717f5aab0a8abc8f7362a91a22d2L45-R45)
* Modified `modules/container-registry/main.tf` to use the `use_zone_redundancy` variable instead of `use_private_networking` for zone redundancy configuration.

***NOTE*** running the `make pre-commit`  `./avm docscheck`  `pre-commit`  `pr-check`  commands made the changes below.  

### Documentation Updates:
* Added descriptions for the `use_zone_redundancy` variable in `README.md` and `modules/container-registry/README.md`, explaining its purpose and default value. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R805-R812) [[2]](diffhunk://#diff-45b149b6e637b7287e1332e3406cb668350fbf871e8c2bf9aa1ed7e664a4c341R218-R225)

### Container Runtime Validation:
* Enhanced the `avm` script to skip container runtime checks if the `AVM_IN_CONTAINER` environment variable is set, improving compatibility for containerized environments.
* Introduced a conditional pull option (`--pull always`) in the `avm` script, controlled by the `AVM_NO_CONTAINER_PULL` environment variable. [[1]](diffhunk://#diff-f786ee298282510219c2b3ec5d83994f14dc6d5bcc6c6579268acdb07a804bb1R24-R30) [[2]](diffhunk://#diff-f786ee298282510219c2b3ec5d83994f14dc6d5bcc6c6579268acdb07a804bb1L31-R39)

### Miscellaneous:
* Removed an outdated comment in `.github/CODEOWNERS` without affecting ownership assignments.

Summary above assisted by GitHub Copilot, and provided troubleshooting support during the discovery phase.

## Type of Change

<!-- Use the check-boxes [x] on the options that are relevant. -->

- [ ] Non-module change (e.g. CI/CD, documentation, etc.)
- [X] Azure Verified Module updates:
  - [X] Bugfix containing backwards compatible bug fixes
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [X] Feature update backwards compatible feature updates.
  - [ ] Breaking changes.
  - [X] Update to documentation

# Checklist

- [X] I'm sure there are no other open Pull Requests for the same update/change
- [X] My corresponding pipelines / checks run clean and green without any errors or warnings
- [X] I did run all  [pre-commit](https://azure.github.io/Azure-Verified-Modules/contributing/terraform/terraform-contribution-flow/#5-run-pre-commit-checks) checks

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/terraform -->

jaredfholgate commented on main.container.app.environment.tf:
```diff
@@ -9,7 +9,7 @@ resource "azurerm_container_app_environment" "this" {
   log_analytics_workspace_id         = local.log_analytics_workspace_id
   tags                               = var.tags
-  zone_redundancy_enabled            = var.use_private_networking ? true : null
+  zone_redundancy_enabled            = var.use_zone_redundancy ? true : false
```
Have you tested this in the following combinations:

- Private networking off / zone redundant off
- Private networking on / zone redundant off
- Private networking off / zone redundant on
- Private networking on / zone redundant off

I strongly suspect it was linked to private networking by design since they are tightly couple by the resource sku deployed with this setting.

johnwildes replied:
@jaredfholgate apologies for the time waste on this one.  It does appear I have missed both the documentation and that test in my attempt to solve the issue I was running into with `northcentralus` deployment.

The tests that do work
- Private networking off / Zone redundant off (pass)
- Private networking on / Zone redundant off (pass)
- Private networking on / Zone redundant off (pass)
- Private networking off / Zone redundant on (fail) as per documentation on CAE provider.

![image](https://github.com/user-attachments/assets/6e731ba7-4797-4ec0-80c3-18d8c569cd6b)

johnwildes replied:
As it currently stands, with no changes to the module, the use of private networking as the variable to control zone redundancy prevents the use of this module in a private manner in 20 regions (per this reference [here](https://www.azurespeed.com/Information/AzureRegions). 

In all cases where we've built runners for customers, they've always wanted them with private networking.  With private networking on and the ability to self-select the zone redundancy, this allows for the edge cases where you must deploy to a region that has no availability zone.  We're opening up access to EVERY zone to deploy these privately, but we assume the risk of outage in the case that the region has DC issues.  Every one of these regions if you put resources here every resource has that issue.

I come to the point where I ask myself, is there a case where I want to deploy these resources publicly and need zone redundancy, and I have none to speak of.  Are there more cases where customers would be comfortable with private networking and no redundancy I would argue there are.

At this point Every Customer in Australia deploying to an Australian Region cannot use this module as is.  Every Candian customer, deploying to Canadian regions cannot use this module as is.  Everywhere else in the world, With this change we allow them to be able to use this module.  Everywhere else in the world has atleast one region where they can deploy this as is.

With this change we do allow every region to deploy this model, in a best practice manner, but we introduce one case where it cannot be used, and that is publicly exposing the ACE and runner environment AND requiring Zone redundancy.

jaredfholgate replied:
To be clear, I wasn't suggesting we don't implement, just add an extra check for private networking for the scenario that doesn't work. 

I agree it is good to have the option to deploy private in a region that does not support AZ.

johnwildes replied:
Thanks @jaredfholgate , your comments above prompted me to mentally walk through the change rather than just focusing on the fact that I couldn't deploy it in `northcentralus`.  I'll work on the check for the scenario that doesn't work.

jaredfholgate commented on main.container.app.environment.tf:
```diff
@@ -9,7 +9,7 @@ resource "azurerm_container_app_environment" "this" {
   log_analytics_workspace_id         = local.log_analytics_workspace_id
   tags                               = var.tags
-  zone_redundancy_enabled            = var.use_private_networking ? true : null
+  zone_redundancy_enabled            = var.use_zone_redundancy ? true : false
```
The ternary is redundant here since the var is a boolean.

jaredfholgate commented on main.container.app.environment.tf:
```diff
@@ -9,7 +9,7 @@ resource "azurerm_container_app_environment" "this" {
   log_analytics_workspace_id         = local.log_analytics_workspace_id
   tags                               = var.tags
-  zone_redundancy_enabled            = var.use_private_networking ? true : null
+  zone_redundancy_enabled            = var.use_zone_redundancy ? true : false
```
Have you tested setting this to false vs not supplying with null, it was clearly done that was by design originally.

johnwildes replied:
updated to null

jaredfholgate commented on main.container.registry.tf:
```diff
@@ -12,6 +12,7 @@ module "container_registry" {
   private_dns_zone_id                     = local.container_registry_dns_zone_id
   subnet_id                               = local.container_registry_private_endpoint_subnet_id
   tags                                    = var.tags
+  use_zone_redundancy                     = var.use_zone_redundancy ? true : false
```
The ternary is redundant since the var is a boolean.

johnwildes replied:
updated

jaredfholgate commented on variables.tf:
```diff
@@ -86,3 +86,9 @@ variable "use_private_networking" {
+variable "use_zone_redundancy" {
+  type        = bool
+  default     = true
+  description = "Whether or not to use zone redundancy for the container app environment."
```
The description is not accurate, since this also applies to the container registry and public IP.

johnwildes replied:
Updated the description

jaredfholgate reviewed (changes requested):
Hi @johnwildes, thanks for submitting this PR, it is very much appreciated.

There a few opportunities to clean up the code here.

The biggest thing we need is validation that the combinations of settings actually work. There was clearly a reason for tightly coupling private networking and zone redundancy in the first instance and the docs specify that the combinations you are now allowing are not allowed.

See the note in the docs here: https://registry.terraform.io/providers/hashicorp/azurerm/4.34.0/docs/resources/container_app_environment#zone_redundancy_enabled-1

<img width="496" alt="image" src="https://github.com/user-attachments/assets/cc4acd5b-dc68-4a87-90bb-8d3dbcd73176" />

johnwildes commented:
@jaredfholgate I've updated here with validation for the use case of `use_private_networking` as `false` and `use_zone_redundancy` as `true` so it will fail plan when this is the case.   I've also added validation for when `use_zone_redundancy` is `true` but the location doesn't support it.  To fail on plan.  I've tested both scenarios and received the appropriate errors when in `terraform plan`

Region validation
![image](https://github.com/user-attachments/assets/b4109f9a-e2b5-464b-b112-2f3d182ee961)

Case - Private Networking off, a Zone redundancy on
![image](https://github.com/user-attachments/assets/2ecc1165-7a13-4d04-92b5-c9ea70509d7e)
